lib/scheduler.pm

```diff
@@ -120,12 +119,14 @@ Parse variables and test modules from a yaml file representing a test suite to b
 
 sub load_yaml_schedule {
     if (my $yamlfile = get_var('YAML_SCHEDULE')) {
-        my $schedule      = $ypp->load_file($root_project_dir . $yamlfile);
-        my %schedule_vars = parse_vars($schedule);
+        my $schedule              = $ypp->load_file($root_project_dir . $yamlfile);
+        my %schedule_vars         = parse_vars($schedule);
+        my $test_context_instance = undef;
         while (my ($var, $value) = each %schedule_vars) { set_var($var, $value) }
         my @schedule_modules = parse_schedule($schedule);
         parse_test_suite_data($schedule);
-        loadtest($_) for (@schedule_modules);
+        $test_context_instance = exists $schedule_vars{TEST_CONTEXT} ? $schedule_vars{TEST_CONTEXT}->new() : get_var('TEST_CONTEXT', '')->new();
```

Comment:
I don't get why we want to make it that complex, with this checks, this code is not any faster than single `get_var` call as it just gets value from the hash.
On top `get_var('TEST_CONTEXT', '')->new();` is very dangerous as if variable is not define you will try to call method on empty string.

Reply:
> I don't get why we want to make it that complex, with this checks, this code is not any faster than single `get_var` call as it just gets value from the hash.
> On top `get_var('TEST_CONTEXT', '')->new();` is very dangerous as if variable is not define you will try to call method on empty string.

+1 for the last part. my mistake is that i had another idea in the beginning and then i added the `->new()` there without check.

so you prefer straight `get_var`. no problem. just remind me what happens if the by any means the get_var has foo and vars in yaml dont? isnt it more save to say if it is not there check outside?

Reply:
It doesn't make sense, we set all the variables from yaml 3 lines above, so `get_var` will have access to the latest consistent state of variables. It's not my personal preference, we just have no reason to make this complication of the code. We loose readability and do not gain any performance.
Also, note the difference between `exists` and `defined`, because if you set value to `undef`, exists returns true, as key is there, and then you again call `new()` on `undef`.

Reply:
One more thing, as you use just string, I would propose to add some validation, see https://stackoverflow.com/questions/433752/how-can-i-determine-if-a-perl-function-exists-at-runtime

So if test dies we could print some suspenseful hint which variable to check, instead of `cannot locate method new()...`.

Reply:
i see now. thanks a lot man. gonna fix it

Reply:
@rwx788 it is updated. feel free to review when the CI pass. reopen it if you see something wrong.